File: rsws_model/src/resource.rs
//! 资源模型

use chrono::{DateTime, Utc};
use salvo_oapi::ToSchema;
use rust_decimal::Decimal;
use serde::{Deserialize, Serialize};
use sqlx::FromRow;

/// 资源归属类型常量
pub const OWNER_TYPE_USER: &str = "user";
pub const OWNER_TYPE_PLATFORM: &str = "platform";

/// 资源
#[derive(Debug, Clone, Serialize, Deserialize, FromRow, ToSchema)]
pub struct Resource {
    pub id: i64,
    pub title: String,
    pub description: Option<String>,
    pub price: Decimal,
    pub category_id: Option<i64>,
    pub file_url: Option<String>,
    pub thumbnail_url: Option<String>,
    pub is_active: bool,
    pub detail_description: Option<String>,
    pub specifications: Option<serde_json::Value>,
    pub usage_guide: Option<String>,
    pub precautions: Option<String>,
    pub display_images: Option<serde_json::Value>,
    pub owner_type: String,
    pub provider_id: Option<i64>,
    pub supported_os: Option<serde_json::Value>,
    pub commission_rate: i64,
    pub download_count: i64,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

/// 创建资源请求
#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
pub struct CreateResourceRequest {
    pub title: String,
    pub description: Option<String>,
    pub price: Decimal,
    pub category_id: Option<i64>,
    pub file_url: Option<String>,
    pub thumbnail_url: Option<String>,
    pub detail_description: Option<String>,
    pub specifications: Option<serde_json::Value>,
    pub usage_guide: Option<String>,
    pub precautions: Option<String>,
    pub display_images: Option<Vec<String>>,
    pub supported_os: Option<Vec<String>>,
}

/// 更新资源请求
#[derive(Debug, Clone, Serialize, Deserialize, Default, ToSchema)]
pub struct UpdateResourceRequest {
    pub title: Option<String>,
    pub description: Option<String>,
    pub price: Option<Decimal>,
    pub category_id: Option<i64>,
    pub file_url: Option<String>,
    pub thumbnail_url: Option<String>,
    pub is_active: Option<bool>,
    pub detail_description: Option<String>,
    pub specifications: Option<serde_json::Value>,
    pub usage_guide: Option<String>,
    pub precautions: Option<String>,
    pub display_images: Option<Vec<String>>,
    pub supported_os: Option<Vec<String>>,
}

/// 资源列表响应
#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
pub struct ResourceListResponse {
    pub items: Vec<Resource>,
    pub total: i64,
    pub page: i64,
    pub page_size: i64,
}

// ==================== 单元测试 ====================

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_create_resource_request() {
        let req = CreateResourceRequest {
            title: "Test Resource".to_string(),
            description: Some("A test resource".to_string()),
            price: Decimal::new(1000, 0),
            category_id: None,
            file_url: None,
            thumbnail_url: None,
            detail_description: None,
            specifications: None,
            usage_guide: None,
            precautions: None,
            display_images: None,
            supported_os: None,
        };

        assert_eq!(req.title, "Test Resource");
        assert_eq!(req.price, Decimal::new(1000, 0));
    }
}

File: ui/user/src/api/resource.ts
import request, { type ApiResponse } from './request'

export interface Resource {
  id: number
  title: string
  description: string | null
  price: number
  category_id: number | null
  thumbnail_url: string | null
  cover_image?: string | null
  is_active: boolean
  detail_description?: string | null
  specifications?: any
  usage_guide?: string | null
  precautions?: string | null
  display_images?: string[]
  download_count: number
  provider_type?: string
  commission_rate?: number
  created_at: string
  updated_at: string
}

export interface ResourceListResponse {
  items: Resource[]
  total: number
  page: number
  page_size: number
  total_pages: number
}

/** 资源详情（购买后有 file_url） */
export type ResourceDetail = Resource & {
  file_url?: string | null
}

export interface PurchaseCheckResponse {
  purchased: boolean
  order_id?: number
}

export interface DownloadInfo {
  file_url: string
  file_name: string
}

/** 获取资源列表（公开端点） */
export async function listResources(params?: {
  page?: number
  page_size?: number
  category_id?: number
  search?: string
}): Promise<ApiResponse<ResourceListResponse>> {
  return request.get('/resource', { params })
}

/** 获取资源详情（公开端点） */
export async function getResource(id: number): Promise<ApiResponse<ResourceDetail>> {
  return request.get(`/resource/${id}`)
}

/** 检查用户是否已购买某资源 */
export async function checkPurchase(resourceId: number): Promise<ApiResponse<PurchaseCheckResponse>> {
  return request.get(`/resource/${resourceId}/purchase-check`)
}

/** 获取资源下载信息（需要已购买） */
export async function getDownloadInfo(resourceId: number): Promise<ApiResponse<DownloadInfo>> {
  return request.get(`/resource/${resourceId}/download`)
}
